cupy/statistics/correlation.py

```diff
@@ -48,7 +59,83 @@ def corrcoef(a, y=None, rowvar=True, bias=None, ddof=None):
     return out
 
 
-# TODO(okuta): Implement correlate
+def correlate(a, v, mode='valid'):
+    """Returns the cross-correlation of two 1-dimensional sequences.
+
+    Args:
+        a (cupy.ndarray): first 1-dimensional input.
+        v (cupy.ndarray): second 1-dimensional input.
+        mode (optional): `valid`, `same`, `full`
+
+    Returns:
+        cupy.ndarray: Discrete cross-correlation of a and v.
+
+    .. seealso:: :func:`numpy.correlate`
+
+    """
+    if a.ndim != 1 or v.ndim != 1:
+        raise ValueError("object too deep for desired array")
+    conj_v = cupy.conj(v[::-1])
+    method = cupyx.scipy.signal.choose_conv_method(a, conj_v, mode)
+    if method == 'direct':
+        if cupy.iscomplexobj(v):
+            v = cupy.conj(v)
+        inverted, out = _dot_correlate(a, v, mode)
+        if inverted:
+            out = out[::-1]
+    elif method == 'fft':
+        out = cupy.math.misc._fftconvolve1d(a, conj_v, mode)
+    else:
+        raise ValueError('Unsupported method')
+    return out
+
+
+def _dot_correlate(a1, a2, mode):
+    inverted = 0
+    dtype = cupy.result_type(a1, a2)
+    if a1.size == 0 or a2.size == 0:
+        raise ValueError("Array arguments cannot be empty")
+    if a1.size < a2.size:
+        a1, a2 = a2, a1
+        inverted = 1
+    length = n1 = a1.size
+    n = n2 = a2.size
+    left, right, length = _generate_boundaries(mode, length, n)
+    output = cupy.zeros(length, dtype)
+    a1 = a1.astype(dtype, copy=False)
+    a2 = a2.astype(dtype, copy=False)
+    start = j1 = 0
+    j2 = left
+    n -= left
+    for i in range(left):
+        dot_kernel(a1[:n], a2[j2:], output[i])
+        n += 1
+        j2 -= 1
```

Comment:
@asi1024  Can this be parallelized ?
I understand that Reduction kernel parallelizes along reduction-axis and also elementwise one. But I am asking about sliding window with each iteration. I thought about unrolling the sliding window over 1D input in a 2D form but this seems inefficient. (Which made me use dynamic parallelism in the first place ...)

Reply:
Create two arrays of shape `(out_size, max(a1.size, a2.size))`, and reduce with the second axis.

Reply:
I don't think I get this. This is what I intended to do
``` python
    n1 = a1.size
    n2 = a2.size
    start = 0
    for i in range(left):
        dot_kernel(a1[:n2 - left + i], a2[left - i:], output[i])
    start += left
    A1 = _rolling_window(a1, n2)
    A2 = cupy.repeat(a2[None, :], n1 - n2 + 1, axis=0)
    dot_kernel(A1, A2, output[start:(start + n1 - n2 + 1)], axis=1)
    start += n1 - n2 + 1
    for i in range(right):
        dot_kernel(a1[n1 - n2 + 1 + i:], a2[:n2 - 1 - i], output[start + i])
```
where n1  >=  n2 and `_rolling_window` is

```python
def _rolling_window(a, window):
    shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
    strides = a.strides + (a.strides[-1],)
    return cupy.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
```